File: clients/dotnet/tests/VectorsTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Xunit;

namespace Quidnug.Client.Tests;

/// <summary>
/// .NET SDK consumer for the v1.0 cross-SDK test vectors at
/// <c>docs/test-vectors/v1.0/</c>.
///
/// Asserts the five conformance properties via the Quidnug.Client
/// public API (Quid.Sign, Quid.Verify, Quid.FromPublicHex,
/// Quid.FromPrivateScalarHex) against the checked-in reference
/// vectors.
///
/// NOTE: the local dev environment at commit time had a broken
/// .NET 9 runtime config, so these tests could not be run
/// locally. Verification in CI is pending. The test file compiles
/// in isolation (only System.Text.Json + Xunit + Quidnug.Client).
/// </summary>
public class VectorsTests
{
    private static string VectorsRoot()
    {
        // Test runs from clients/dotnet/tests/bin/Debug/net8.0/ typically.
        // Walk up until we find docs/test-vectors/v1.0.
        DirectoryInfo? dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir != null)
        {
            var candidate = Path.Combine(dir.FullName, "docs", "test-vectors", "v1.0");
            if (Directory.Exists(candidate)) return candidate;
            dir = dir.Parent;
        }
        throw new InvalidOperationException(
            "docs/test-vectors/v1.0 not found walking up from " + AppContext.BaseDirectory);
    }

    private static Dictionary<string, JsonElement> LoadKeys()
    {
        var result = new Dictionary<string, JsonElement>();
        foreach (var path in Directory.GetFiles(
                Path.Combine(VectorsRoot(), "test-keys"), "*.json"))
        {
            var doc = JsonDocument.Parse(File.ReadAllText(path));
            result[doc.RootElement.GetProperty("name").GetString()!] = doc.RootElement.Clone();
        }
        return result;
    }

    private static JsonElement LoadVectorFile(string name)
    {
        var doc = JsonDocument.Parse(File.ReadAllText(
            Path.Combine(VectorsRoot(), name)));
        return doc.RootElement.Clone();
    }

    private static string BytesToHex(byte[] b)
    {
        var sb = new StringBuilder(b.Length * 2);
        foreach (var byt in b) sb.Append(byt.ToString("x2"));
        return sb.ToString();
    }

    private static byte[] HexToBytes(string hex)
    {
        var b = new byte[hex.Length / 2];
        for (int i = 0; i < b.Length; i++)
        {
            b[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
        }
        return b;
    }

    /// <summary>
    /// Run the conformance properties for a single case.
    /// </summary>
    private static void RunCase(JsonElement caseNode, JsonElement key,
            byte[] signable, string derivedId)
    {
        var expected = caseNode.GetProperty("expected");
        var name = caseNode.GetProperty("name").GetString()!;

        // Property 1: SHA-256 matches.
        byte[] hash = SHA256.HashData(signable);
        Assert.Equal(expected.GetProperty("sha256_of_canonical_hex").GetString(),
                BytesToHex(hash));

        // Property 2: hex and utf8 forms match.
        Assert.Equal(expected.GetProperty("canonical_signable_bytes_hex").GetString(),
                BytesToHex(signable));
        Assert.Equal(expected.GetProperty("canonical_signable_bytes_utf8").GetString(),
                Encoding.UTF8.GetString(signable));

        // Property 3: ID derivation.
        Assert.Equal(expected.GetProperty("expected_id").GetString(), derivedId);

        // Property 4: reference signature verifies.
        using var qRO = Quid.FromPublicHex(
                key.GetProperty("public_key_sec1_hex").GetString()!);
        Assert.True(qRO.Verify(signable,
                expected.GetProperty("reference_signature_hex").GetString()!),
            $"{name}: SDK Verify rejected reference signature");
        Assert.Equal(64, expected.GetProperty("signature_length_bytes").GetInt32());

        // Property 5: tampered signature rejects.
        byte[] tampered = HexToBytes(expected.GetProperty("reference_signature_hex").GetString()!);
        tampered[5] ^= 0x01;
        Assert.False(qRO.Verify(signable, BytesToHex(tampered)),
            $"{name}: tampered signature accepted");

        // Property 6: SDK sign-verify round-trip.
        using var qSign = Quid.FromPrivateScalarHex(
                key.GetProperty("private_scalar_hex").GetString()!);
        string sdkSig = qSign.Sign(signable);
        Assert.Equal(128, sdkSig.Length);  // 64 bytes = 128 hex
        Assert.True(qSign.Verify(signable, sdkSig));
    }

    // --- Per-type signable-bytes builders (Go-struct-order JSON) ---
    //
    // System.Text.Json writes dictionary entries in insertion order
    // (unlike Newtonsoft); we exploit this by adding fields in the
    // order the Go struct declares them.

    private static byte[] TrustSignable(JsonElement inp)
    {
        using var ms = new MemoryStream();
        using (var writer = new Utf8JsonWriter(ms, new JsonWriterOptions { SkipValidation = false }))
        {
            writer.WriteStartObject();
            writer.WriteString("id", inp.GetProperty("id").GetString());
            writer.WriteString("type", "TRUST");
            writer.WriteString("trustDomain", inp.GetProperty("trustDomain").GetString());
            writer.WriteNumber("timestamp", inp.GetProperty("timestamp").GetInt64());
            writer.WriteString("signature", "");
            writer.WriteString("publicKey", inp.GetProperty("publicKey").GetString());
            writer.WriteString("truster", inp.GetProperty("truster").GetString());
            writer.WriteString("trustee", inp.GetProperty("trustee").GetString());
            WriteGoCompatFloat(writer, "trustLevel", inp.GetProperty("trustLevel").GetDouble());
            writer.WriteNumber("nonce", inp.GetProperty("nonce").GetInt64());
            if (TryGetNonEmptyString(inp, "description", out var desc))
                writer.WriteString("description", desc);
            if (TryGetNonZeroInt64(inp, "validUntil", out var vu))
                writer.WriteNumber("validUntil", vu);
            writer.WriteEndObject();
        }
        return ms.ToArray();
    }

    private static string TrustId(JsonElement inp)
    {
        using var ms = new MemoryStream();
        using (var writer = new Utf8JsonWriter(ms))
        {
            writer.WriteStartObject();
            writer.WriteString("Truster", inp.GetProperty("truster").GetString());
            writer.WriteString("Trustee", inp.GetProperty("trustee").GetString());
            WriteGoCompatFloat(writer, "TrustLevel", inp.GetProperty("trustLevel").GetDouble());
            writer.WriteString("TrustDomain", inp.GetProperty("trustDomain").GetString());
            writer.WriteNumber("Timestamp", inp.GetProperty("timestamp").GetInt64());
            writer.WriteEndObject();
        }
        return BytesToHex(SHA256.HashData(ms.ToArray()));
    }

    private static byte[] IdentitySignable(JsonElement inp)
    {
        using var ms = new MemoryStream();
        using (var writer = new Utf8JsonWriter(ms))
        {
            writer.WriteStartObject();
            writer.WriteString("id", inp.GetProperty("id").GetString());
            writer.WriteString("type", "IDENTITY");
            writer.WriteString("trustDomain", inp.GetProperty("trustDomain").GetString());
            writer.WriteNumber("timestamp", inp.GetProperty("timestamp").GetInt64());
            writer.WriteString("signature", "");
            writer.WriteString("publicKey", inp.GetProperty("publicKey").GetString());
            writer.WriteString("quidId", inp.GetProperty("quidId").GetString());
            writer.WriteString("name", inp.GetProperty("name").GetString());
            if (TryGetNonEmptyString(inp, "description", out var desc))
                writer.WriteString("description", desc);
            if (inp.TryGetProperty("attributes", out var attrs) && attrs.ValueKind == JsonValueKind.Object)
            {
                writer.WritePropertyName("attributes");
                attrs.WriteTo(writer);
            }
            writer.WriteString("creator", inp.GetProperty("creator").GetString());
            writer.WriteNumber("updateNonce", inp.GetProperty("updateNonce").GetInt64());
            if (TryGetNonEmptyString(inp, "homeDomain", out var hd))
                writer.WriteString("homeDomain", hd);
            writer.WriteEndObject();
        }
        return ms.ToArray();
    }

    private static string IdentityId(JsonElement inp)
    {
        using var ms = new MemoryStream();
        using (var writer = new Utf8JsonWriter(ms))
        {
            writer.WriteStartObject();
            writer.WriteString("QuidID", inp.GetProperty("quidId").GetString());
            writer.WriteString("Name", inp.GetProperty("name").GetString());
            writer.WriteString("Creator", inp.GetProperty("creator").GetString());
            writer.WriteString("TrustDomain", inp.GetProperty("trustDomain").GetString());
            writer.WriteNumber("UpdateNonce", inp.GetProperty("updateNonce").GetInt64());
            writer.WriteNumber("Timestamp", inp.GetProperty("timestamp").GetInt64());
            writer.WriteEndObject();
        }
        return BytesToHex(SHA256.HashData(ms.ToArray()));
    }

    private static void WriteGoCompatFloat(Utf8JsonWriter writer, string name, double value)
    {
        // Go's encoding/json elides ".0" on integer-valued floats.
        // Mirror that: whole-number floats in int64 range go as integers.
        if (double.IsFinite(value) && value == Math.Floor(value) && Math.Abs(value) < 1e15)
            writer.WriteNumber(name, (long)value);
        else
            writer.WriteNumber(name, value);
    }

    private static bool TryGetNonEmptyString(JsonElement node, string name, out string value)
    {
        if (node.TryGetProperty(name, out var p) && p.ValueKind == JsonValueKind.String)
        {
            value = p.GetString()!;
            return !string.IsNullOrEmpty(value);
        }
        value = "";
        return false;
    }

    private static bool TryGetNonZeroInt64(JsonElement node, string name, out long value)
    {
        if (node.TryGetProperty(name, out var p) && p.ValueKind == JsonValueKind.Number)
        {
            value = p.GetInt64();
            return value != 0;
        }
        value = 0;
        return false;
    }

    // --- Test cases ---

    [Fact]
    public void TrustVectors()
    {
        var vf = LoadVectorFile("trust-tx.json");
        Assert.Equal("TRUST", vf.GetProperty("tx_type").GetString());
        var keys = LoadKeys();
        foreach (var c in vf.GetProperty("cases").EnumerateArray())
        {
            var key = keys[c.GetProperty("signer_key_ref").GetString()!];
            var inp = c.GetProperty("input");
            byte[] signable = TrustSignable(inp);
            string id = TrustId(inp);
            RunCase(c, key, signable, id);
        }
    }

    [Fact]
    public void IdentityVectors()
    {
        var vf = LoadVectorFile("identity-tx.json");
        Assert.Equal("IDENTITY", vf.GetProperty("tx_type").GetString());
        var keys = LoadKeys();
        foreach (var c in vf.GetProperty("cases").EnumerateArray())
        {
            var key = keys[c.GetProperty("signer_key_ref").GetString()!];
            var inp = c.GetProperty("input");
            byte[] signable = IdentitySignable(inp);
            string id = IdentityId(inp);
            RunCase(c, key, signable, id);
        }
    }

    [Fact]
    public void SdkSignProducesIEEE1363()
    {
        using var q = Quid.Generate();
        byte[] data = Encoding.UTF8.GetBytes("test-data");
        string sigHex = q.Sign(data);
        Assert.Equal(128, sigHex.Length);
        Assert.True(q.Verify(data, sigHex));
    }
}

File: clients/swift/Tests/QuidnugTests/VectorsTests.swift
import XCTest
import CryptoKit
@testable import Quidnug

/// Swift SDK consumer for the v1.0 cross-SDK test vectors at
/// `docs/test-vectors/v1.0/`.
///
/// Asserts the five conformance properties via the Swift SDK's
/// public API (`Quid.sign`, `Quid.verify`, `Quid.fromPublicHex`,
/// `Quid.fromPrivateScalarHex`) against the checked-in reference
/// vectors.
///
/// NOTE: Swift isn't available on the author's dev box (Windows
/// only), so these tests could not be executed locally. The file
/// compiles against Swift 5.9+ + CryptoKit + Foundation as a
/// logical continuation of the existing test suite. Verification
/// in macOS / Linux CI is pending.
final class VectorsTests: XCTestCase {

    private static func vectorsRoot() throws -> URL {
        // Walk up from the test bundle / executable path until we
        // find docs/test-vectors/v1.0.
        let fm = FileManager.default
        var url = URL(fileURLWithPath: fm.currentDirectoryPath)
        for _ in 0..<10 {
            let candidate = url.appendingPathComponent("docs")
                .appendingPathComponent("test-vectors")
                .appendingPathComponent("v1.0")
            if fm.fileExists(atPath: candidate.path) {
                return candidate
            }
            url = url.deletingLastPathComponent()
        }
        // Fall back to relative path from the test source file.
        let thisFile = URL(fileURLWithPath: #filePath)
        return thisFile
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("docs")
            .appendingPathComponent("test-vectors")
            .appendingPathComponent("v1.0")
    }

    struct TestKey: Codable {
        let name: String
        let seed: String
        let privateScalarHex: String
        let publicKeySEC1Hex: String
        let quidId: String

        enum CodingKeys: String, CodingKey {
            case name, seed
            case privateScalarHex = "private_scalar_hex"
            case publicKeySEC1Hex = "public_key_sec1_hex"
            case quidId = "quid_id"
        }
    }

    private static func loadKeys() throws -> [String: TestKey] {
        let dir = try vectorsRoot().appendingPathComponent("test-keys")
        let contents = try FileManager.default.contentsOfDirectory(atPath: dir.path)
        var out: [String: TestKey] = [:]
        for name in contents where name.hasSuffix(".json") {
            let data = try Data(contentsOf: dir.appendingPathComponent(name))
            let k = try JSONDecoder().decode(TestKey.self, from: data)
            out[k.name] = k
        }
        return out
    }

    private static func loadVectorFile(_ name: String) throws -> [String: Any] {
        let url = try vectorsRoot().appendingPathComponent(name)
        let data = try Data(contentsOf: url)
        let obj = try JSONSerialization.jsonObject(with: data)
        guard let dict = obj as? [String: Any] else {
            throw QuidnugError.validation("\(name): not a JSON object")
        }
        return dict
    }

    private static func bytesToHex(_ b: Data) -> String {
        return b.map { String(format: "%02x", $0) }.joined()
    }

    private static func hexToBytes(_ s: String) -> Data {
        var out = Data()
        var i = s.startIndex
        while i < s.endIndex {
            let j = s.index(i, offsetBy: 2)
            if let byte = UInt8(s[i..<j], radix: 16) {
                out.append(byte)
            }
            i = j
        }
        return out
    }

    /// Run all five conformance properties for a single case.
    private func runCase(
        caseObj: [String: Any],
        key: TestKey,
        signable: Data,
        derivedId: String
    ) throws {
        let name = caseObj["name"] as? String ?? "<no-name>"
        guard let expected = caseObj["expected"] as? [String: Any] else {
            XCTFail("\(name): missing expected"); return
        }

        // Property 1: SHA-256 matches.
        let digest = SHA256.hash(data: signable)
        let shaHex = Self.bytesToHex(Data(digest))
        XCTAssertEqual(
            shaHex,
            expected["sha256_of_canonical_hex"] as? String,
            "\(name): SHA-256 mismatch"
        )

        // Property 2: hex + utf8 canonical forms match.
        XCTAssertEqual(
            Self.bytesToHex(signable),
            expected["canonical_signable_bytes_hex"] as? String,
            "\(name): canonical hex diverges"
        )
        XCTAssertEqual(
            String(data: signable, encoding: .utf8),
            expected["canonical_signable_bytes_utf8"] as? String,
            "\(name): canonical utf8 diverges"
        )

        // Property 3: ID derivation.
        XCTAssertEqual(
            derivedId,
            expected["expected_id"] as? String,
            "\(name): ID mismatch"
        )

        // Property 4: reference signature verifies via SDK.
        let qRO = try Quid.fromPublicHex(key.publicKeySEC1Hex)
        let refSigHex = expected["reference_signature_hex"] as! String
        XCTAssertTrue(
            qRO.verify(signable, sigHex: refSigHex),
            "\(name): SDK Verify rejected reference signature"
        )
        XCTAssertEqual(
            (expected["signature_length_bytes"] as? Int) ?? -1,
            64,
            "\(name): signature_length_bytes != 64"
        )

        // Property 5: tampered signature rejects.
        var tampered = Self.hexToBytes(refSigHex)
        tampered[5] ^= 0x01
        XCTAssertFalse(
            qRO.verify(signable, sigHex: Self.bytesToHex(tampered)),
            "\(name): tampered signature accepted"
        )

        // Property 6: SDK sign-verify round-trip.
        let qSign = try Quid.fromPrivateScalarHex(key.privateScalarHex)
        let sdkSig = try qSign.sign(signable)
        XCTAssertEqual(sdkSig.count, 128, "\(name): SDK signature not 64 bytes")
        XCTAssertTrue(qSign.verify(signable, sigHex: sdkSig),
            "\(name): SDK sign-verify round-trip failed")
    }

    // --- Per-type signable-bytes builders (Go struct order) ---

    private static func trustSignable(_ inp: [String: Any]) -> Data {
        var pairs: [(String, Any)] = [
            ("id", inp["id"] as? String ?? ""),
            ("type", "TRUST"),
            ("trustDomain", inp["trustDomain"] as? String ?? ""),
            ("timestamp", inp["timestamp"] as? Int ?? 0),
            ("signature", ""),
            ("publicKey", inp["publicKey"] as? String ?? ""),
            ("truster", inp["truster"] as? String ?? ""),
            ("trustee", inp["trustee"] as? String ?? ""),
            ("trustLevel", goCompatFloat(inp["trustLevel"] as? Double ?? 0)),
            ("nonce", inp["nonce"] as? Int ?? 0),
        ]
        if let desc = inp["description"] as? String, !desc.isEmpty {
            pairs.append(("description", desc))
        }
        if let vu = inp["validUntil"] as? Int, vu != 0 {
            pairs.append(("validUntil", vu))
        }
        return encodePairs(pairs)
    }

    private static func trustId(_ inp: [String: Any]) -> String {
        let pairs: [(String, Any)] = [
            ("Truster", inp["truster"] as? String ?? ""),
            ("Trustee", inp["trustee"] as? String ?? ""),
            ("TrustLevel", goCompatFloat(inp["trustLevel"] as? Double ?? 0)),
            ("TrustDomain", inp["trustDomain"] as? String ?? ""),
            ("Timestamp", inp["timestamp"] as? Int ?? 0),
        ]
        let data = encodePairs(pairs)
        let sha = SHA256.hash(data: data)
        return bytesToHex(Data(sha))
    }

    private static func identitySignable(_ inp: [String: Any]) -> Data {
        var pairs: [(String, Any)] = [
            ("id", inp["id"] as? String ?? ""),
            ("type", "IDENTITY"),
            ("trustDomain", inp["trustDomain"] as? String ?? ""),
            ("timestamp", inp["timestamp"] as? Int ?? 0),
            ("signature", ""),
            ("publicKey", inp["publicKey"] as? String ?? ""),
            ("quidId", inp["quidId"] as? String ?? ""),
            ("name", inp["name"] as? String ?? ""),
        ]
        if let desc = inp["description"] as? String, !desc.isEmpty {
            pairs.append(("description", desc))
        }
        if let attrs = inp["attributes"] as? [String: Any], !attrs.isEmpty {
            pairs.append(("attributes", attrs))
        }
        pairs.append(("creator", inp["creator"] as? String ?? ""))
        pairs.append(("updateNonce", inp["updateNonce"] as? Int ?? 0))
        if let hd = inp["homeDomain"] as? String, !hd.isEmpty {
            pairs.append(("homeDomain", hd))
        }
        return encodePairs(pairs)
    }

    private static func identityId(_ inp: [String: Any]) -> String {
        let pairs: [(String, Any)] = [
            ("QuidID", inp["quidId"] as? String ?? ""),
            ("Name", inp["name"] as? String ?? ""),
            ("Creator", inp["creator"] as? String ?? ""),
            ("TrustDomain", inp["trustDomain"] as? String ?? ""),
            ("UpdateNonce", inp["updateNonce"] as? Int ?? 0),
            ("Timestamp", inp["timestamp"] as? Int ?? 0),
        ]
        let data = encodePairs(pairs)
        return bytesToHex(Data(SHA256.hash(data: data)))
    }

    /// Match Go's `encoding/json` behavior of eliding ".0" on
    /// integer-valued floats.
    private static func goCompatFloat(_ v: Double) -> Any {
        if v.isFinite, v.rounded() == v, abs(v) < 1e15 {
            return Int(v)
        }
        return v
    }

    /// Serialize ordered key-value pairs into struct-declaration-order
    /// JSON, matching Go's `json.Marshal` on a typed struct.
    ///
    /// We use JSONSerialization on an ordered NSDictionary-like
    /// structure via manual JSON assembly to guarantee order.
    /// (Swift's JSONSerialization sorts keys when .sortedKeys is
    /// set but otherwise iterates an unordered dict — so we build
    /// the JSON string by hand.)
    private static func encodePairs(_ pairs: [(String, Any)]) -> Data {
        var json = "{"
        for (i, (k, v)) in pairs.enumerated() {
            if i > 0 { json += "," }
            json += "\"\(k)\":\(encodeValue(v))"
        }
        json += "}"
        return Data(json.utf8)
    }

    private static func encodeValue(_ v: Any) -> String {
        if let s = v as? String {
            let encoded = s
                .replacingOccurrences(of: "\\", with: "\\\\")
                .replacingOccurrences(of: "\"", with: "\\\"")
                .replacingOccurrences(of: "\n", with: "\\n")
                .replacingOccurrences(of: "\r", with: "\\r")
                .replacingOccurrences(of: "\t", with: "\\t")
            return "\"\(encoded)\""
        }
        if let i = v as? Int {
            return String(i)
        }
        if let d = v as? Double {
            // Non-integer-valued float path; rare in v1.0 vectors.
            return String(d)
        }
        if let dict = v as? [String: Any] {
            // Objects within payload/attributes: use JSONSerialization
            // with .sortedKeys for stability of nested keys (matches
            // Go's map[string]interface{} alphabetization).
            if let data = try? JSONSerialization.data(
                withJSONObject: dict,
                options: [.sortedKeys, .withoutEscapingSlashes]),
               let s = String(data: data, encoding: .utf8) {
                return s
            }
            return "{}"
        }
        if let arr = v as? [Any] {
            let parts = arr.map { encodeValue($0) }
            return "[\(parts.joined(separator: ","))]"
        }
        if let b = v as? Bool {
            return b ? "true" : "false"
        }
        return "null"
    }

    // --- Tests ---

    func testTrustVectors() throws {
        let keys = try Self.loadKeys()
        let vf = try Self.loadVectorFile("trust-tx.json")
        XCTAssertEqual(vf["tx_type"] as? String, "TRUST")
        let cases = vf["cases"] as! [[String: Any]]
        XCTAssertFalse(cases.isEmpty)

        for c in cases {
            let ref = c["signer_key_ref"] as! String
            let key = keys[ref]!
            let inp = c["input"] as! [String: Any]
            let signable = Self.trustSignable(inp)
            let id = Self.trustId(inp)
            try runCase(caseObj: c, key: key, signable: signable, derivedId: id)
        }
    }

    func testIdentityVectors() throws {
        let keys = try Self.loadKeys()
        let vf = try Self.loadVectorFile("identity-tx.json")
        XCTAssertEqual(vf["tx_type"] as? String, "IDENTITY")
        let cases = vf["cases"] as! [[String: Any]]
        XCTAssertFalse(cases.isEmpty)

        for c in cases {
            let ref = c["signer_key_ref"] as! String
            let key = keys[ref]!
            let inp = c["input"] as! [String: Any]
            let signable = Self.identitySignable(inp)
            let id = Self.identityId(inp)
            try runCase(caseObj: c, key: key, signable: signable, derivedId: id)
        }
    }

    func testSdkSignProducesIEEE1363() throws {
        let q = Quid.generate()
        let data = Data("test-data".utf8)
        let sigHex = try q.sign(data)
        XCTAssertEqual(sigHex.count, 128,
            "v1.0 mandates 64-byte IEEE-1363 (128 hex chars)")
        XCTAssertTrue(q.verify(data, sigHex: sigHex))
    }
}
